pkg/cache/status.go

```diff
@@ -0,0 +1,138 @@
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package cache
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/envoyproxy/go-control-plane/api"
+)
+
+// NodeID provides the function to obtain a primary key from the node.
+type NodeID func(*api.Node) string
+
+// DefaultID is the default ID function combining the node cluster and the node ID.
+func DefaultID(node *api.Node) string {
+	return fmt.Sprintf("%s/%s", node.Cluster, node.Id)
+}
+
+// VersionCompare compares version identifiers and returns of three values, +1 if
+// if first argument is greater than the second, 0 if equal, and -1 if less.
+type VersionCompare func(string, string) int
+
+// NodeInfo is node status.
+type NodeInfo struct {
+	id string
+	mu sync.RWMutex
+
+	// latest acknowledged version
+	acked map[ResponseType]string
+
+	// latest responded version
+	responded map[ResponseType]string
+}
+
+// StatusInfo is a status tracker for configuration version acknowledgement by
+// individual proxy nodes.
+type StatusInfo struct {
+	id      NodeID
+	compare VersionCompare
+
+	mu    sync.RWMutex
+	nodes map[string]*NodeInfo
+}
+
+// NewStatusInfo creates a new status tracker.
+func NewStatusInfo(id NodeID, compare VersionCompare) *StatusInfo {
+	return &StatusInfo{
+		id:      id,
+		compare: compare,
+		nodes:   make(map[string]*NodeInfo),
+	}
+}
+
+// GetInfo retrieves node info or creates a record for the node.
+func (info *StatusInfo) GetInfo(node *api.Node) *NodeInfo {
+	info.mu.Lock()
+	defer info.mu.Unlock()
+
+	id := info.id(node)
+	ni, exists := info.nodes[id]
+	if !exists {
+		ni = &NodeInfo{
+			id:        id,
+			acked:     make(map[ResponseType]string),
+			responded: make(map[ResponseType]string),
+		}
+		info.nodes[id] = ni
+	}
+	return ni
+}
+
+// Ack records an acknowledgement for the latest version from a remote node.
+func (info *StatusInfo) Ack(typ ResponseType, node *api.Node, version string) {
+	ni := info.GetInfo(node)
+
+	ni.mu.Lock()
+	defer ni.mu.Unlock()
+	if old, exists := ni.acked[typ]; !exists || info.compare(version, old) > 0 {
+		ni.acked[typ] = version
+	}
+}
+
+// AllowResponse checks whether a response should be allowed..
+func (info *StatusInfo) AllowResponse(typ ResponseType, node *api.Node, version string) bool {
+	ni := info.GetInfo(node)
+
+	ni.mu.Lock()
+	defer ni.mu.Unlock()
+	if ni.responded[typ] == version {
+		// reject if the version is responded once already
+		// there are two main causes of rejection by the proxy:
+		// - internal validity is violated:
+		//   the response fails to validate and repeated responses do not lead to progress
+		// - referential validity is violated (e.g. validate_clusters option):
+		//   retries may lead to progress but this requires Envoy to enforce referential validity (a non-default behavior)
+		//   it is important to stagger updates and not immediately reply in this case
+		return false
+	}
+
+	ni.responded[typ] = version
```

Comment:
If this is the only place where this is being updated, do we even need a lock ? I guess what I am driving at is whether this code is in the fast path (though it’s funny to say fast path in control plane code)